{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 93,
   "id": "8e0570ce",
   "metadata": {},
   "outputs": [],
   "source": [
    "import os\n",
    "import math\n",
    "import random\n",
    "import itertools as it\n",
    "import numpy as np\n",
    "import concurrent.futures as cf\n",
    "from scipy import stats\n",
    "from scipy.stats.mstats import gmean\n",
    "from statsmodels.stats import multitest as mt\n",
    "from sklearn.decomposition import PCA as skPCA\n",
    "pca = skPCA(n_components=3)\n",
    "import matplotlib.pyplot as plt\n",
    "import matplotlib.mlab as mlab\n",
    "import matplotlib.patches as mpatches\n",
    "from matplotlib.backends.backend_pdf import PdfPages\n",
    "import seaborn as sns\n",
    "plt.rcParams['pdf.fonttype'] = 42\n",
    "plt.rcParams['ps.fonttype'] = 42\n",
    "sns.set()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "36f0701a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def mysplit(c):\n",
    "    tl = c.lstrip('0123456789')\n",
    "    hd = c[:len(c)-len(tl)]\n",
    "    return(hd,tl)\n",
    "def readMatrix(fpath,offset=False,iglines = 0,split=False):\n",
    "    dicto = {}\n",
    "    with open(fpath,'r') as f:\n",
    "        allc = [l.strip().split('\\t') for l in f.readlines()]\n",
    "    #if '*' in allc[0][0]:\n",
    "    #    hdr = [tuple(e.split('*')) for e in allc[0]]\n",
    "    #else:\n",
    "    #    hdr = [tuple(e.split('_')) for e in allc[0]]\n",
    "    allc = allc[iglines:]\n",
    "    hdr = allc[0]\n",
    "    if split:\n",
    "        hdr = [mysplit(e[1:]) for e in hdr]\n",
    "    if offset:\n",
    "        hdr = hdr[1:]\n",
    "    dicto ={h:{} for h in hdr}\n",
    "    for l in allc[1:]:\n",
    "        for i,h in enumerate(hdr):\n",
    "            dicto[h][l[0]] = l[i+1]\n",
    "        #for l in allc[1:]:\n",
    "    #    dicto[l[0]] = {hdr[i]:l[i+1] for i in range(len(hdr))}\n",
    "    return dicto"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a347fc8e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def sampleCounts(dicto,firstN):\n",
    "    bigList = []\n",
    "    names=[]\n",
    "    counts = []\n",
    "    if sum(dicto.values())<firstN:\n",
    "            return None\n",
    "    for g in dicto:\n",
    "        for j in range(dicto[g]):\n",
    "            bigList.append(g)\n",
    "        #names.append(r)\n",
    "        #counts.append(dicto[r])\n",
    "    #for i,count in enumerate(counts):\n",
    "    #    for j in range(count):\n",
    "    #        bigList.append(names[i])\n",
    "    #if len(bigList)<firstN:\n",
    "    #    return None\n",
    "    if len(bigList)!=sum(dicto.values()):\n",
    "        print(\"MISMATCH\")\n",
    "    random.shuffle(bigList)\n",
    "    subList = bigList[:firstN]\n",
    "    retDict = {k:0 for k in dicto}\n",
    "    for elt in subList:\n",
    "        retDict[elt] += 1\n",
    "    return retDict"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "03f54e00",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plotPCA(ax,counts,samples,groups,colors,anno=False):\n",
    "    res = pca.fit_transform(counts)\n",
    "    print(len(res))\n",
    "    for i,s in enumerate(samples):\n",
    "        ax.scatter(res[i][0], res[i][1],color=colors[groups[i]],alpha=0.75,edgecolors='none')#,s=100)#plt.cm.spectral(cond_idx[conditions[i]]), alpha=0.7)\n",
    "    if anno:\n",
    "        for i,s in enumerate(samples):\n",
    "            ax.annotate(s, xy = (res[i][0], res[i][1]), xytext = (4, 4), textcoords = 'offset points')\n",
    "    \n",
    "    handles = []\n",
    "    for grp in set(groups):\n",
    "        handles.append(mpatches.Patch(color=colors[grp],label=grp))\n",
    "    \n",
    "    expvar = pca.explained_variance_ratio_\n",
    "    ax.set_xlabel('PC1: {:.2f}% variance explained'.format(100*expvar[0]))\n",
    "    ax.set_ylabel('PC2: {:.2f}% variance explained'.format(100*expvar[1]))\n",
    "    \n",
    "    return handles\n",
    "    #ax.legend(handles=handles,prop={'size':8},loc='best')\n",
    "    #plt.tight_layout()\n",
    "    #plt.show()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "id": "dea4893b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Directory definitions\n",
    "#indir: contains count data, e.g. from FeatureCounts, for all samples (matrix for riboseq/rnaseq same as for ribo/PLATEseq)\n",
    "indir = '../data/'\n",
    "#outdir: output directory for figures\n",
    "outdir = '../figs/'\n",
    "#wDir: working directory containing DESeq2 results, normalized counts ranked lists, gene sets, and other text files \n",
    "wDir = '../workingdir/'\n",
    "#gsDir: working directory containing GSEA output\n",
    "gsdir = '../gsea/'\n",
    "#refDir: source directory for genomic reference information not provided in working directory\n",
    "refDir = '../ReferenceInfo/'\n",
    "#gspath: path to GSEA CLI\n",
    "gspath = '/home/ubuntu/datadrive/riboPLATEseq/GSEA_4.1.0/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "18c48e2b",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "464\n"
     ]
    }
   ],
   "source": [
    "#Non-polyadenylated genes; mostly identifies histones, snoRNAs, and scaRNAs, among other ncRNA\n",
    "#From Yang et al 2011, \"Genomewide characterization of non-polyadenylated RNAs\" (https://doi.org/10.1186/gb-2011-12-2-r16)\n",
    "#These genes were originally identified with symbols from hg19 (UCSC \"Known Genes\" predictor); conversion of unrecognized \n",
    "#loci to hg38-recognized GENCODE symbols required searching NCBI Nucleotide for ~100 source records\n",
    "with open(refDir+\"genes_non_pA.txt\",'r') as f:\n",
    "    badG = [e.strip() for e in f.readlines()]\n",
    "print(len(badG))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "id": "60894a32",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "13611 22085\n",
      "10334 3277\n"
     ]
    }
   ],
   "source": [
    "#Non-coding gene set determined from RefSeq GTF annotation\n",
    "ncGenes = set()\n",
    "cGenes = set()\n",
    "\n",
    "with open(refDir+\"GRCh38_refseq_ERCC.gtf\",'r') as f:\n",
    "    for line in f:\n",
    "        try:\n",
    "            gtfE = line.strip().split('\\t')[-1].strip().split(';')\n",
    "            nmDict[gtfE[1].strip().split(' ')[1].strip('\"').split('.')[0]]=gtfE[0].strip().split(' ')[1].strip('\"')\n",
    "            ent = line.split('\\t')[-1].split(';')\n",
    "            elt = ent[1].split()[1]\n",
    "            if 'NR_' in elt:\n",
    "                ncGenes.add(ent[0].split()[1].strip('\"'))\n",
    "            elif 'NM_' in elt:\n",
    "                cGenes.add(ent[0].split()[1].strip('\"'))\n",
    "        except Exception:\n",
    "            pass\n",
    "print(len(ncGenes),len(cGenes))\n",
    "qcGenes = ncGenes & cGenes\n",
    "ncGenes = ncGenes - cGenes\n",
    "\n",
    "print(len(ncGenes),len(qcGenes))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "e840bf95",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2222 309\n",
      "2070 291\n"
     ]
    }
   ],
   "source": [
    "#TOP gene candidate transcripts (NCBI RefSeq; \"NM_\") from Yamashita et al. 2011, \"Comprehensive detection of human terminal\n",
    "#oligo-pyrimidine (TOP) genes and analysis of their characteristics\" (doi: https://dx.doi.org/10.1093%2Fnar%2Fgkn248)\n",
    "#Specific subset of candidate transcripts chosen have homologues in the mouse genome that also contain TOP motifs\n",
    "with open(refDir+\"candidateTOP.txt\",'r') as f:\n",
    "    allC = [l.split('\\t') for l in f.readlines()[1:]]\n",
    "    topC = [l[0].split(',') for l in allC]\n",
    "    tpc = []\n",
    "    for c in topC:\n",
    "        for ci in c:\n",
    "            tpc.append(ci)\n",
    "    tpc = list(set(tpc))\n",
    "    \n",
    "    topm = [l[0].split(',') for l in allC if l[6]!='0\\n']\n",
    "    tpcm = []\n",
    "    for c in topm:\n",
    "        for ci in c:\n",
    "            tpcm.append(ci)\n",
    "    tpcm = list(set(tpcm))\n",
    "print(len(tpc),len(tpcm))\n",
    "\n",
    "#Classical TOP gene set, defined in other studies but also annotated in Yamashita et al\n",
    "with open(refDir+\"classicalTOP.txt\",'r') as f:\n",
    "    classicTOP = [l.strip() for l in f.readlines()]\n",
    "print(len([t for t in tpc if t in nmDict]),len([t for t in tpcm if t in nmDict]))\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "4203deae",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "PLATE 96\n",
      "96\n",
      "riboPLATE 96\n",
      "96\n",
      "PLATE_ERCC 96\n",
      "96\n",
      "riboPLATE_ERCC 96\n",
      "96\n"
     ]
    }
   ],
   "source": [
    "barcode_files = ['PLATE','riboPLATE','PLATE_ERCC','riboPLATE_ERCC']\n",
    "bcDict = {}\n",
    "for bc_pfx in barcode_files:\n",
    "    bc_fl = 'barcodes_{}.tsv'.format(bc_pfx)\n",
    "    with open(indir+bc_fl,'r') as f:\n",
    "        f.readline()\n",
    "        allbc = [l.strip().split('\\t') for l in f.readlines()]\n",
    "    print(bc_pfx,len(allbc))\n",
    "    bcDict[bc_pfx] = {int(l[0]):(l[1],l[2]) for l in allbc}\n",
    "    print(len(bcDict[bc_pfx]))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "id": "f1a3c284",
   "metadata": {},
   "outputs": [],
   "source": [
    "#maplist: vectorized representation of conditions in riboPLATE-seq study by barcode index\n",
    "maplist = ['AZD', 'CTRL', 'MNK1', 'CTRL', 'PP242', 'CTRL', 'BKM', 'MNK1+BKM', 'CTRL', 'CTRL', 'MNK1+BKM', '4EGI1', \n",
    "           'CTRL', 'AZD', 'CTRL', 'MNK1', 'CTRL', 'PP242', 'CTRL', 'BKM', 'MNK1+BKM', 'MNK1+BKM', 'CTRL', 'CTRL', \n",
    "           'CTRL', 'CTRL', 'AZD', 'CTRL', 'MNK1', 'CTRL', 'PP242', 'CTRL', 'BKM', 'CTRL', 'MNK1+BKM', 'MNK1+BKM', \n",
    "           'PP242+MNK1', 'CTRL', 'CTRL', 'AZD', 'CTRL', 'MNK1', 'CTRL', 'PP242', 'CTRL', 'BKM', 'CTRL', 'CTRL',\n",
    "           'CTRL', 'PP242+MNK1', 'PP242+MNK1', 'CTRL', 'AZD', 'CTRL', 'MNK1', 'CTRL', 'PP242', 'CTRL', 'BKM', '4EGI1', \n",
    "           'PP242+BKM', 'CTRL', 'CTRL', 'PP242+MNK1', 'PP242+MNK1', 'AZD', 'CTRL', 'MNK1', 'CTRL', 'PP242', 'CTRL', 'BKM',\n",
    "           'CTRL', 'PP242+BKM', 'PP242+BKM', 'CTRL', '4EGI1', 'PP242+MNK1', 'CTRL', '4EGI1', 'CTRL', '4EGI1', 'CTRL', '4EGI1',\n",
    "           'CTRL', 'CTRL', 'CTRL', 'PP242+BKM', 'PP242+BKM', 'CTRL', 'CTRL', 'PP242+BKM', 'CTRL', 'CTRL', 'CTRL', 'CTRL']"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "id": "832f2bbe",
   "metadata": {},
   "outputs": [],
   "source": [
    "#rDict: readDict; holds {sample:{gene:counts for gene in genes} for sample in samples} \n",
    "#for each set of samples pooled in each (ribo)PLATE library\n",
    "#barcodes are converted to numeric indices (0-95) for ease of reading/addressing individual wells\n",
    "rDict = {}\n",
    "#cdDict: holds {index:sample condition} for each (ribo)PLATE library \n",
    "cdDict = {}\n",
    "for ct in ['riboPLATE','PLATE','riboPLATE_ERCC','PLATE_ERCC']:\n",
    "    tmp = readMatrix(indir+ct+'.counts.txt')\n",
    "    rDict[ct] = {k:{g:int(v) for (g,v) in tmp[k].items()} for k in tmp}\n",
    "    bcs = bcDict[ct]\n",
    "    bc_idx = {bcs[b][0]:b for b in bcs}\n",
    "    rDict[ct] = {bc_idx[k]:rDict[ct][k] for k in rDict[ct]}\n",
    "    #index to condition for each barcode in study\n",
    "    cdDict[ct] = {bc:bcs[bc][1] for bc in bcs}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "5ab203cc",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "10\n",
      "23\n"
     ]
    }
   ],
   "source": [
    "fullseq = {}\n",
    "for lbl in ['RPF','RNA']:\n",
    "    tmp = readMatrix(indir+lbl+'.counts.txt')\n",
    "    fullseq.update({tuple(k.split('.')):{g:int(v) for (g,v) in tmp[k].items()} for k in tmp})\n",
    "    print(len(fullseq))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "id": "7ddc3260",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "riboPLATE.final\n",
      "riboPLATE.final.cts.txt riboPLATE.final.cols.txt\n",
      "85 170\n"
     ]
    }
   ],
   "source": [
    "#Setup for FINAL DESeq2 on riboPLATE -- excludes outliers, excludes ERCC\n",
    "wDir = '/home/ubuntu/datadrive/riboPLATEseq/riboPLATEseq_repo/workingdir/'\n",
    "badS = [0,4,6,9,10,11,22,34,36,60,76]\n",
    "\n",
    "sampleIdx = [i for i in range(96)]\n",
    "bc_rb = bcDict['riboPLATE']#[bcDict['riboPLATE'][i][0] for i in sampleIdx]\n",
    "bc_rn = bcDict['PLATE']#[bcDict['PLATE'][i][0] for i in sampleIdx]\n",
    "conditions = list(set(maplist))\n",
    "conditions = [''.join(e.split('+')) for e in conditions]\n",
    "conditionDict = {c:[i for i in range(96) if ''.join(bc_rb[i][1].split('+'))==c and ''.join(bc_rn[i][1].split('+'))==c] for c in conditions}#                    for [i for i in range(96) if ''.join(maplist[i].split('+'))==c] for c in conditions}\n",
    "rb = 'riboPLATE'\n",
    "rn = 'PLATE'\n",
    "gsub = [g for g in goodG if 'ERCC-' not in g]\n",
    "#gSums_rb = {g:sum([rDict[rb][i][g] for i in rDict[rb]]) for g in gsub}\n",
    "#gSums_rn =  {g:sum([rDict[rn][i][g] for i in rDict[rn]]) for g in gsub}\n",
    "lbl = 'final'\n",
    "pfx = rb+'.'+lbl\n",
    "print(pfx)\n",
    "clFl = '{}.cols.txt'.format(pfx)\n",
    "ctFl = '{}.cts.txt'.format(pfx)\n",
    "samples = []\n",
    "hdr = []\n",
    "print(ctFl,clFl)\n",
    "with open(wDir+ctFl,'w') as o, open(wDir+clFl,'w') as col:\n",
    "    col.write('condition,type\\n')\n",
    "    for cond in conditions:\n",
    "        for s in conditionDict[cond]:\n",
    "            if s in badS: continue\n",
    "            col.write(\"ribo_{},{},RIBO\\n\".format(s,cond))\n",
    "            hdr.append(\"ribo_{}\".format(s))\n",
    "            samples.append(s)\n",
    "    for cond in conditions:\n",
    "        for s in conditionDict[cond]:\n",
    "            if s in badS: continue\n",
    "            col.write(\"rna_{},{},RNA\\n\".format(s,cond))\n",
    "            hdr.append(\"rna_{}\".format(s))\n",
    "\n",
    "    print(len(samples),len(hdr))\n",
    "    o.write('\\t'.join(hdr)+'\\n')\n",
    "    #o.write(\"\\t\".join(['ribo_{}'.format(s) for s in samples])+'\\n')\n",
    "    for g in gsub:\n",
    "        o.write(\"{}\\t{}\\t{}\\n\".format(g,'\\t'.join([str(int(rDict[rb][i][g])) for i in samples]),'\\t'.join([str(int(rDict[rn][i][g])) for i in samples])))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "id": "7f8d9185",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "riboPLATE.all\n",
      "riboPLATE.all.cts.txt riboPLATE.all.cols.txt\n",
      "96 192\n"
     ]
    }
   ],
   "source": [
    "#Setup for DESeq2 on riboPLATE with outliers\n",
    "#wDir = '/home/ubuntu/datadrive/riboPLATEseq/riboPLATEseq_repo/workingdir/'\n",
    "\n",
    "sampleIdx = [i for i in range(96)]\n",
    "bc_rb = bcDict['riboPLATE']#[bcDict['riboPLATE'][i][0] for i in sampleIdx]\n",
    "bc_rn = bcDict['PLATE']#[bcDict['PLATE'][i][0] for i in sampleIdx]\n",
    "conditions = list(set(maplist))\n",
    "conditions = [''.join(e.split('+')) for e in conditions]\n",
    "conditionDict = {c:[i for i in range(96) if ''.join(bc_rb[i][1].split('+'))==c and ''.join(bc_rn[i][1].split('+'))==c] for c in conditions}#                    for [i for i in range(96) if ''.join(maplist[i].split('+'))==c] for c in conditions}\n",
    "rb = 'riboPLATE'\n",
    "rn = 'PLATE'\n",
    "gsub = [g for g in goodG if 'ERCC-' not in g]\n",
    "#gSums_rb = {g:sum([rDict[rb][i][g] for i in rDict[rb]]) for g in gsub}\n",
    "#gSums_rn =  {g:sum([rDict[rn][i][g] for i in rDict[rn]]) for g in gsub}\n",
    "lbl = 'all'\n",
    "pfx = rb+'.'+lbl\n",
    "print(pfx)\n",
    "clFl = '{}.cols.txt'.format(pfx)\n",
    "ctFl = '{}.cts.txt'.format(pfx)\n",
    "samples = []\n",
    "hdr = []\n",
    "print(ctFl,clFl)\n",
    "with open(wDir+ctFl,'w') as o, open(wDir+clFl,'w') as col:\n",
    "    col.write('condition,type\\n')\n",
    "    for cond in conditions:\n",
    "        for s in conditionDict[cond]:\n",
    "            col.write(\"ribo_{},{},RIBO\\n\".format(s,cond))\n",
    "            hdr.append(\"ribo_{}\".format(s))\n",
    "            samples.append(s)\n",
    "    for cond in conditions:\n",
    "        for s in conditionDict[cond]:\n",
    "            col.write(\"rna_{},{},RNA\\n\".format(s,cond))\n",
    "            hdr.append(\"rna_{}\".format(s))\n",
    "\n",
    "    print(len(samples),len(hdr))\n",
    "    o.write('\\t'.join(hdr)+'\\n')\n",
    "    for g in gsub:\n",
    "        o.write(\"{}\\t{}\\t{}\\n\".format(g,'\\t'.join([str(int(rDict[rb][i][g])) for i in samples]),'\\t'.join([str(int(rDict[rn][i][g])) for i in samples])))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "id": "6bbd8513",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "riboPLATE_ERCC.final\n",
      "riboPLATE_ERCC.final.cts.txt riboPLATE_ERCC.final.cols.txt\n",
      "96 192\n"
     ]
    }
   ],
   "source": [
    "#Setup for DESeq2 on ERCC dataset\n",
    "#wDir = '/home/ubuntu/datadrive/riboPLATEseq/riboPLATEseq_repo/workingdir/'\n",
    "\n",
    "sampleIdx = [i for i in range(96)]\n",
    "bc_rb = bcDict['riboPLATE_ERCC']#[bcDict['riboPLATE'][i][0] for i in sampleIdx]\n",
    "bc_rn = bcDict['PLATE_ERCC']#[bcDict['PLATE'][i][0] for i in sampleIdx]\n",
    "conditions = list(set(cdDict['PLATE_ERCC'].values()))\n",
    "conditions = [''.join(e.split('+')) for e in conditions]\n",
    "conditionDict = {c:[i for i in range(96) if ''.join(bc_rb[i][1].split('+'))==c and ''.join(bc_rn[i][1].split('+'))==c] for c in conditions}#                    for [i for i in range(96) if ''.join(maplist[i].split('+'))==c] for c in conditions}\n",
    "rb = 'riboPLATE_ERCC'\n",
    "rn = 'PLATE_ERCC'\n",
    "gsub = [g for g in goodG]\n",
    "lbl = 'final'\n",
    "pfx = rb+'.'+lbl\n",
    "print(pfx)\n",
    "clFl = '{}.cols.txt'.format(pfx)\n",
    "ctFl = '{}.cts.txt'.format(pfx)\n",
    "samples = []\n",
    "hdr = []\n",
    "print(ctFl,clFl)\n",
    "with open(wDir+ctFl,'w') as o, open(wDir+clFl,'w') as col:\n",
    "    col.write('ERCC,type\\n')\n",
    "    for cond in conditions:\n",
    "        for s in conditionDict[cond]:\n",
    "            col.write(\"ribo_{},{},RIBO\\n\".format(s,cond))\n",
    "            hdr.append(\"ribo_{}\".format(s))\n",
    "            samples.append(s)\n",
    "    for cond in conditions:\n",
    "        for s in conditionDict[cond]:\n",
    "            col.write(\"rna_{},{},RNA\\n\".format(s,cond))\n",
    "            hdr.append(\"rna_{}\".format(s))\n",
    "\n",
    "    print(len(samples),len(hdr))\n",
    "    o.write('\\t'.join(hdr)+'\\n')\n",
    "    #o.write(\"\\t\".join(['ribo_{}'.format(s) for s in samples])+'\\n')\n",
    "    for g in gsub:\n",
    "        o.write(\"{}\\t{}\\t{}\\n\".format(g,'\\t'.join([str(int(rDict[rb][i][g])) for i in samples]),'\\t'.join([str(int(rDict[rn][i][g])) for i in samples])))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "2cf2a661",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('DMSO_30_1', 'RIBO'), ('DMSO_30_2', 'RIBO'), ('DMSO_30_3', 'RIBO'), ('PP_30_1', 'RIBO'), ('PP_30_2', 'RIBO'), ('PP_30_3', 'RIBO'), ('PP_6_1', 'RIBO'), ('PP_6_3', 'RIBO'), ('DMSO_6_3', 'RIBO'), ('DMSO_6_4', 'RIBO'), ('DMSO_30_1', 'RNA'), ('DMSO_30_2', 'RNA'), ('DMSO_30_3', 'RNA'), ('PP_30_1', 'RNA'), ('PP_30_2', 'RNA'), ('PP_30_3', 'RNA'), ('DMSO_6_1', 'RNA'), ('DMSO_6_2', 'RNA'), ('DMSO_6_3', 'RNA'), ('PP_6_1', 'RNA'), ('PP_6_2', 'RNA'), ('PP_6_3', 'RNA'), ('DMSO_6_4', 'RNA')]\n",
      "['DMSO_6', 'DMSO_30', 'PP_30', 'PP_6']\n",
      "{'DMSO_6': [('DMSO_6_3', 'RIBO'), ('DMSO_6_4', 'RIBO'), ('DMSO_6_1', 'RNA'), ('DMSO_6_2', 'RNA'), ('DMSO_6_3', 'RNA'), ('DMSO_6_4', 'RNA')], 'DMSO_30': [('DMSO_30_1', 'RIBO'), ('DMSO_30_2', 'RIBO'), ('DMSO_30_3', 'RIBO'), ('DMSO_30_1', 'RNA'), ('DMSO_30_2', 'RNA'), ('DMSO_30_3', 'RNA')], 'PP_30': [('PP_30_1', 'RIBO'), ('PP_30_2', 'RIBO'), ('PP_30_3', 'RIBO'), ('PP_30_1', 'RNA'), ('PP_30_2', 'RNA'), ('PP_30_3', 'RNA')], 'PP_6': [('PP_6_1', 'RIBO'), ('PP_6_3', 'RIBO'), ('PP_6_1', 'RNA'), ('PP_6_2', 'RNA'), ('PP_6_3', 'RNA')]}\n"
     ]
    }
   ],
   "source": [
    "#Setup for DESeq2 on riboseq/RNAseq dataset: excluding ERCC\n",
    "samples = [e for e in fullseq]\n",
    "conditions = list(set(e[0][:-2] for e in samples))\n",
    "print(samples)\n",
    "print(conditions)\n",
    "conditionDict = {c:[s for s in samples if s[0][:-2]==c] for c in conditions}\n",
    "print(conditionDict)\n",
    "conditionsTS = conditions\n",
    "\n",
    "ctFl = 'FL.final.cts.txt'\n",
    "clFl = 'FL.final.cols.txt'\n",
    "\n",
    "ggs = [g for g in fullseq[samples[0]].keys() if g in goodG and 'ERCC-' not in g]\n",
    "lbl = '.final'\n",
    "with open(wDir+ctFl,'w') as o, open(wDir+clFl,'w') as col:\n",
    "    col.write('condition,type\\n')\n",
    "    hdr = []\n",
    "    samples = []\n",
    "    for c in conditions:\n",
    "        for s in conditionDict[c]:\n",
    "            col.write('{},{},{}\\n'.format('_'.join(s),c,s[1]))\n",
    "            samples.append(s)\n",
    "            hdr.append('_'.join(s))\n",
    "    o.write('\\t'.join(hdr)+'\\n')\n",
    "    for g in ggs:\n",
    "        o.write(\"{}\\t{}\\n\".format(g,'\\t'.join([str(int(fullseq[s][g])) for s in samples])))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "id": "17e91765",
   "metadata": {},
   "outputs": [
    {
     "ename": "SyntaxError",
     "evalue": "invalid syntax (2068712720.py, line 1)",
     "output_type": "error",
     "traceback": [
      "\u001b[0;36m  File \u001b[0;32m\"/tmp/ipykernel_783991/2068712720.py\"\u001b[0;36m, line \u001b[0;32m1\u001b[0m\n\u001b[0;31m    cmd = Rscript deseq_xform.R ../workingdir/ riboPLATE.final ctint ~condition+type+condition:type\u001b[0m\n\u001b[0m                  ^\u001b[0m\n\u001b[0;31mSyntaxError\u001b[0m\u001b[0;31m:\u001b[0m invalid syntax\n"
     ]
    }
   ],
   "source": [
    "cmd = \"Rscript deseq_xform.R {} riboPLATE.final ctint ~condition+type+condition:type\".format(wDir)\n",
    "os.system(cmd)\n",
    "cmd = \"Rscript deseq_xform_2b.R {} FL.final ctint ~condition+type+condition:type\".format(wDir)\n",
    "os.system(cmd)\n",
    "cmd = \"Rscript deseq_xform.R {} riboPLATE_ERCC.final tbasic ~type\".format(wDir)\n",
    "os.system(cmd)\n",
    "\n",
    "cmd = \"Rscript deseq_xform.R {} riboPLATE.final tbasic ~type\".format(wDir)\n",
    "os.system(cmd)\n",
    "cmd = \"Rscript deseq_xform_2b.R {} FL.final tbasic ~type\".format(wDir)\n",
    "os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "id": "de6b18e7",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Process ForkProcess-97:\n",
      "Process ForkProcess-106:\n",
      "Process ForkProcess-104:\n",
      "Process ForkProcess-102:\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "Traceback (most recent call last):\n",
      "Traceback (most recent call last):\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/concurrent/futures/process.py\", line 237, in _process_worker\n",
      "    call_item = call_queue.get(block=True)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 315, in _bootstrap\n",
      "    self.run()\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/queues.py\", line 102, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/process.py\", line 108, in run\n",
      "    self._target(*self._args, **self._kwargs)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/concurrent/futures/process.py\", line 237, in _process_worker\n",
      "    call_item = call_queue.get(block=True)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/concurrent/futures/process.py\", line 237, in _process_worker\n",
      "    call_item = call_queue.get(block=True)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/concurrent/futures/process.py\", line 237, in _process_worker\n",
      "    call_item = call_queue.get(block=True)\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/queues.py\", line 103, in get\n",
      "    res = self._recv_bytes()\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/queues.py\", line 102, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/queues.py\", line 102, in get\n",
      "    with self._rlock:\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/connection.py\", line 221, in recv_bytes\n",
      "    buf = self._recv_bytes(maxlength)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/connection.py\", line 419, in _recv_bytes\n",
      "    buf = self._recv(4)\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/synchronize.py\", line 95, in __enter__\n",
      "    return self._semlock.__enter__()\n",
      "KeyboardInterrupt\n",
      "  File \"/home/ubuntu/miniconda3/lib/python3.9/multiprocessing/connection.py\", line 384, in _recv\n",
      "    chunk = read(handle, remaining)\n",
      "KeyboardInterrupt\n",
      "KeyboardInterrupt\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "\u001b[0;32m/tmp/ipykernel_783991/2637249734.py\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[1;32m     20\u001b[0m         \u001b[0;32mfor\u001b[0m \u001b[0mk\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mfullseq\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     21\u001b[0m             \u001b[0mftDict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mexecutor\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0msubmit\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0msampleCounts\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0mfullseq\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mk\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m,\u001b[0m\u001b[0msampleNumber\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m]\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mk\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m---> 22\u001b[0;31m         \u001b[0;32mfor\u001b[0m \u001b[0mft\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mcf\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mas_completed\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mftDict\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m     23\u001b[0m             \u001b[0mres\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mft\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mresult\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m     24\u001b[0m             \u001b[0mk\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mftDict\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mft\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/concurrent/futures/_base.py\u001b[0m in \u001b[0;36mas_completed\u001b[0;34m(fs, timeout)\u001b[0m\n\u001b[1;32m    243\u001b[0m                             len(pending), total_futures))\n\u001b[1;32m    244\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 245\u001b[0;31m             \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mevent\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mwait_timeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    246\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    247\u001b[0m             \u001b[0;32mwith\u001b[0m \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mlock\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    572\u001b[0m             \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_flag\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    573\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0;32mnot\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 574\u001b[0;31m                 \u001b[0msignaled\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mself\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0m_cond\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0mwait\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mtimeout\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    575\u001b[0m             \u001b[0;32mreturn\u001b[0m \u001b[0msignaled\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    576\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;32m~/miniconda3/lib/python3.9/threading.py\u001b[0m in \u001b[0;36mwait\u001b[0;34m(self, timeout)\u001b[0m\n\u001b[1;32m    310\u001b[0m         \u001b[0;32mtry\u001b[0m\u001b[0;34m:\u001b[0m    \u001b[0;31m# restore state no matter what (e.g., KeyboardInterrupt)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    311\u001b[0m             \u001b[0;32mif\u001b[0m \u001b[0mtimeout\u001b[0m \u001b[0;32mis\u001b[0m \u001b[0;32mNone\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m--> 312\u001b[0;31m                 \u001b[0mwaiter\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0macquire\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m    313\u001b[0m                 \u001b[0mgotit\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0;32mTrue\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m    314\u001b[0m             \u001b[0;32melse\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "#SATURATION ANALYSIS\n",
    "#runs concurrently, takes a long time\n",
    "\n",
    "tsSamp = [k for k in fullseq]\n",
    "IntervalNumber = 50\n",
    "detG = []\n",
    "sampleR = np.linspace(1e5,1e7,num=IntervalNumber)#np.logspace(4,7,num=IntervalNumber)\n",
    "#t_S: vector of all sample IDs\n",
    "#t_S = [(k,rb) for rb in ['riboPLATE','PLATE','riboPLATE_ERCC','PLATE_ERCC'] for k in range(96)]+[(e[1],e[0]) for e in fullseq]\n",
    "for sampleNumber in [int(e) for e in sampleR]:\n",
    "    ftDict = {}\n",
    "    tmp = {}\n",
    "    #print(int(i))\n",
    "    #dicts = [{g:v for (g,v) in rDict[rb][k].items() if g in goodG} for rb in ['riboPLATE','PLATE','riboPLATE_ERCC','PLATE_ERCC'] for k in rDict[rb]]+[{g:v for (g,v) in fullseq[k].items() if g in goodG} for k in fullseq]# if 'DMSO' in k[0] or 'PP' in k[0]]\n",
    "    #print(len(dicts))\n",
    "    with cf.ProcessPoolExecutor() as executor:\n",
    "        for rb in ['riboPLATE','PLATE','riboPLATE_ERCC','PLATE_ERCC']:\n",
    "            for k in rDict[rb]:\n",
    "                ftDict[executor.submit(sampleCounts,rDict[rb][k],sampleNumber)] = (rb,k)\n",
    "        for k in fullseq:\n",
    "            ftDict[executor.submit(sampleCounts,fullseq[k],sampleNumber)] = k\n",
    "        for ft in cf.as_completed(ftDict):\n",
    "            res = ft.result()\n",
    "            k = ftDict[ft]\n",
    "            if res != None:\n",
    "                tmp[k]=len([g for g in res if res[g]>0])\n",
    "            else:\n",
    "                tmp[k]=0\n",
    "        detG.append(tmp)\n",
    "        #dicts0 = {rb:{k:{g:v for (g,v) in rDict[rb][k].items() if g in goodG} for k in rDict[rb]} for rb in ['riboPLATE','PLATE','riboPLATE_ERCC','PLATE_ERCC']}# for k in rDict[rb]]\n",
    "        #dictGen = it.chain((rDict[r][k] for r in ['riboPLATE','PLATE','riboPLATE_ERCC','PLATE_ERCC'] for k in rDict[r]),(fullseq[k] for k in fullseq))\n",
    "        #1/0\n",
    "        #sampleRB = [e for e in executor.map(sampleCounts,zip(dicts,[i for _ in range(len(dicts))]))]\n",
    "        #sampleRB = executor.map(sampleCounts,dicts,(sampleNumber for  _ in range(len(dicts))))\n",
    "        #tmp = {t_S[j]:len([g for g in srb if srb[g]>0]) if srb!=None else 0 for j,srb in enumerate(sampleRB)}\n",
    "        #detG.append(tmp)        \n",
    "        #tmp = {}\n",
    "        #for j,srb in enumerate(sampleRB):\n",
    "        #    if srb!=None:\n",
    "        #        tmp[t_S[j]]= len([g for g in srb if srb[g]>0])\n",
    "        #    else:\n",
    "        #        tmp[t_S[j]] = 0\n",
    "        #detG.append(tmp)\n",
    "        \n",
    "with open(\"{}SATURATION_OUT.txt\".format(wDir),'w') as o:\n",
    "    o.write('\\t'.join([str(int(i)) for i in sampleR])+'\\n')\n",
    "    for rb in ['riboPLATE','PLATE','riboPLATE_ERCC','PLATE_ERCC']:\n",
    "        for s in rDict[rb]:\n",
    "            o.write(\"{}\\t{}\\n\".format('.'.join([rb,s]),'\\t'.join([str(detG[i][(rb,s)]) for i in range(len(sampleR))])))\n",
    "    for k in fullseq:\n",
    "        o.write(\"{}\\t{}\\n\".format('.'.join(k),'\\t'.join([str(detG[i][k]) for i in range(len(sampleR))])))        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "01f80841",
   "metadata": {},
   "outputs": [],
   "source": [
    "#LIBRARY SATURATION PLOTS\n",
    "#first run \n",
    "satfile = '{}SATURATION_OUT.txt'.format(wDir)\n",
    "satDict = {}\n",
    "with open(satfile,'r') as f:\n",
    "    hdr = [int(e) for e in f.readline().strip().split()]\n",
    "    for line in f:\n",
    "        e = line.split()\n",
    "        satDict[e[0]] = {hdr[i]:int(e[i+1]) for i in range(len(hdr))}\n",
    "\n",
    "rpf = [e for e in satDict if 'RIBO' in e]\n",
    "rna = [e for e in satDict if 'RNA' in e]\n",
    "riboplate = [e for e in satDict if 'riboPLATE_ERCC' in e]\n",
    "rnaplate = [e for e in satDict if 'PLATE_ERCC' in e]\n",
    "riboplate2 = [e for e in satDict if 'riboPLATE' in e and e not in riboplate]\n",
    "rnaplate2 = [e for e in satDict if 'PLATE' in e and e not in rnaplate]\n",
    "\n",
    "for sset,lbl in zip((rpf,rna,riboplate,rnaplate,riboplate2,rnaplate2),('Ribosome Profiling','RNA-seq','riboPLATE-seq_ERCC','PLATE-seq_ERCC','riboPLATE-seq','PLATE-seq')):\n",
    "    pp = PdfPages(\"{}saturation_{}.pdf\".format(outdir,'_'.join(lbl)))\n",
    "    plt.figure(figsize=(10,10))\n",
    "    print(len(sset),len(hdr))\n",
    "    if lbl=='Ribosome Profiling':\n",
    "        plt.xlim(0,5e6)\n",
    "    elif lbl=='RNA-seq':\n",
    "        plt.xlim(0,8e6)\n",
    "    elif lbl == 'riboPLATE-seq':\n",
    "        plt.xlim(0,1.2e6)\n",
    "    elif lbl == 'PLATE-seq':\n",
    "        plt.xlim(0,1e6)\n",
    "        \n",
    "    elif lbl == 'riboPLATE-seq_ERCC':\n",
    "        plt.xlim(0,1.5e5)#200000)\n",
    "    elif lbl == 'PLATE-seq_ERCC':\n",
    "        plt.xlim(0,1e6)\n",
    "\n",
    "    for e in sset:\n",
    "        validR = [v for v in hdr if satDict[e][v]>0]\n",
    "        #print(len(validR))\n",
    "        if len(validR)==0:\n",
    "            continue\n",
    "        if 'ERCC' not in lbl:\n",
    "            validR = [validR[i] for i in (0,4,7)]+validR[10:]\n",
    "        else:\n",
    "            validR = [validR[i] for i in [0]]+validR[2:]\n",
    "\n",
    "        sns.scatterplot(x=[v for v in validR],\n",
    "                        y=[satDict[e][v] for v in validR],\n",
    "                        color='black',\n",
    "                        edgecolor='none'\n",
    "                       )\n",
    "    plt.ylabel(\"Unique Genes Detected\")\n",
    "    plt.xlabel(\"Sampling Depth (Counts Sampled)\")\n",
    "    plt.title(lbl)\n",
    "    pp.savefig()\n",
    "    plt.show()\n",
    "    pp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "841beeed",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Overall plots of genes vs counts, for all libraries (excluding outliers)\n",
    "badS = [0,11,36,60,76]\n",
    "pp = PdfPages(\"{}Complexity.pdf\".format(outdir))\n",
    "plt.figure(figsize=(9,9))\n",
    "plt.title(\"All libraries\")\n",
    "for run in ['riboPLATE','riboseq']:    \n",
    "    if run=='riboPLATE':\n",
    "        rbDict = rDict[run]\n",
    "        rnDict = rDict['PLATE']\n",
    "        smps = range(96)\n",
    "    else:\n",
    "        rbDict = {k[0]:v for (k,v) in fullseq.items() if k[1]=='RIBO'}\n",
    "        rnDict = {k[0]:v for (k,v) in fullseq.items() if k[1]=='RNA'}\n",
    "    ssmp = [s for s in rbDict if s in rnDict]\n",
    "    if run == 'riboPLATE':\n",
    "        rbSum = {s:sum([rbDict[s][v] for v in rbDict[s] if v not in badG and 'ERCC-' not in v]) for s in ssmp}\n",
    "        rnSum = {s:sum([rnDict[s][v] for v in rnDict[s] if v not in badG and 'ERCC-' not in v]) for s in ssmp}\n",
    "        medRD = [np.mean([e for e in rbSum.values()]),np.mean([e for e in rnSum.values()])]\n",
    "    gs = [e for e in rbDict[ssmp[0]].keys() if e not in badG and 'ERCC-' not in e]\n",
    "    print(len(rbDict),len(rnDict),len(ssmp))\n",
    "    plt.scatter([sum([rbDict[s][g] for g in gs]) for s in ssmp],[len([g for g in rbDict[s] if g in gs and rbDict[s][g]>1]) for s in ssmp],label='{} ribo'.format(run),alpha=0.6,edgecolor='none')\n",
    "    plt.scatter([sum([rnDict[s][g] for g in gs]) for s in ssmp],[len([g for g in rnDict[s] if g in gs and rnDict[s][g]>1]) for s in ssmp],label='{} rna'.format(run),alpha=0.6,edgecolor='none')\n",
    "rbDict = {k[0]:{g:v for (g,v) in fullseq[k].items() if g not in badG and 'ERCC-' not in g} for k in fullseq}\n",
    "rnDict = {k[0]:{g:v for (g,v) in fullseq[k].items() if g not in badG and 'ERCC-' not in g} for k in fullseq}\n",
    "rbds = {k:sampleCounts(rbDict[k],int(medRD[0])) for k in rbDict}\n",
    "rnds = {k:sampleCounts(rnDict[k],int(medRD[1])) for k in rnDict}\n",
    "plt.scatter([sum([rbds[s][g] for g in rbds[s]]) for s in ssmp],[len([g for g in rbds[s] if rbds[s][g]>1]) for s in ssmp],label='{} ribo downsampled'.format(run),alpha=0.6,edgecolor='none')\n",
    "plt.scatter([sum([rnds[s][g] for g in rnds[s]]) for s in ssmp],[len([g for g in rnds[s] if rnds[s][g]>1]) for s in ssmp],label='{} rna downsampled'.format(run),alpha=0.6,edgecolor='none')\n",
    "\n",
    "plt.xlabel(\"Sum counts\")\n",
    "plt.ylabel(\"Genes detected\")\n",
    "plt.gca().set_xscale('log')\n",
    "plt.legend()\n",
    "pp.savefig()\n",
    "plt.show()\n",
    "pp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "804a5227",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Load DESeq2 results\n",
    "resDict = {}\n",
    "sigGs = [set(),set()]\n",
    "mod = 'ctint'\n",
    "for c in ['PP242','MNK1','4EGI1','AZD','BKM','MNK1BKM','PP242BKM','PP242MNK1']:\n",
    "    fl = '{od}riboPLATE.final.{md}.condition{c}.typeRIBO.{c}.results.txt'.format(od=wDir,md=mod,c=c)#oDir+'.final.'+mod+'condition{}'\n",
    "    if not os.path.exists(fl):\n",
    "        print(fl)\n",
    "        print(c,\" NOT FOUND\")\n",
    "    else:\n",
    "        with open(fl,'r') as f:\n",
    "            allres = [l.split() for l in f.readlines()[1:]]\n",
    "            resDict[c] = {l[0]:((float(l[2]),float(l[-1])) if 'NA'!=l[-1] else (float(l[2]),1.0)) for l in allres if l[2]!='NA'}    \n",
    "        sigGs[0] |= set([g for g in resDict[c] if resDict[c][g][1]<0.05 and resDict[c][g][0]<0]) \n",
    "        sigGs[1] |= set([g for g in resDict[c] if resDict[c][g][1]<0.05 and resDict[c][g][0]>0]) \n",
    "print(len(sigGs[0]),len(sigGs[1]))\n",
    "\n",
    "for cpfx in ['_6','30']:\n",
    "    fl = '{od}FL.final.{md}.conditionPP{pf}.typeRIBO.DMSO{pf}.results.txt'.format(od=wDir,md=mod,pf=cpfx)\n",
    "    if not os.path.exists(fl):\n",
    "        print(fl)\n",
    "        print('PP'+c,\" NOT FOUND\")\n",
    "    else:\n",
    "        with open(fl,'r') as f:\n",
    "            allres = [l.split() for l in f.readlines()[1:]]\n",
    "            resDict['PP'+c] = {l[0]:((float(l[2]),float(l[-1])) if 'NA'!=l[-1] else (float(l[2]),1.0)) for l in allres if l[2]!='NA'}    \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1c341255",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GSEA setup: genesets (on p-value and direction) AND rankedlists (on effect size)\n",
    "tid='riboPLATE'\n",
    "setDict = {}\n",
    "setDict_excl = {}\n",
    "rankDict = {}\n",
    "cc = ['PP242','MNK1','4EGI1','BKM','AZD']#'MNK1BKM','PP242BKM','PP242MNK1']\n",
    "combos = ['MNK1BKM','PP242BKM','PP242MNK1']\n",
    "for r in cc+combos:\n",
    "    setDict[r] = ([g for g in resDict[r] if resDict[r][g][1]<0.05 and resDict[r][g][0]<0],[g for g in resDict[r] if resDict[r][g][1]<0.05 and resDict[r][g][0]>0])\n",
    "    rankDict[r] = {g:resDict[r][g][0] for g in resDict[r]}\n",
    "    print(r,len(setDict[r][0]),len(setDict[r][1]))\n",
    "#exclusive sets are only exclusive w/in single- and double-treated sets\n",
    "for r in cc:\n",
    "    setDict_excl[r] = ([g for g in setDict[r][0] if len([_ for r in cc if g in setDict[r][0]])==1],[g for g in setDict[r][1] if len([_ for r in cc if g in setDict[r][1]])==1])\n",
    "    print(r,len(setDict_excl[r][0]),len(setDict_excl[r][1]))\n",
    "for r in combos:\n",
    "    setDict_excl[r] = ([g for g in setDict[r][0] if len([_ for r in combos if g in setDict[r][0]])==1],[g for g in setDict[r][1] if len([_ for r in combos if g in setDict[r][1]])==1])\n",
    "    print(r,len(setDict_excl[r][0]),len(setDict_excl[r][1]))\n",
    "\n",
    "setsAcrossCombos = [(set(setDict[c][0]),set(setDict[c][1])) for c in combos]\n",
    "setsInit = setsAcrossCombos[0]\n",
    "for c in range(len(combos)):\n",
    "    setsInit = (setsInit[0]&setsAcrossCombos[c][0],setsInit[1]&setsAcrossCombos[c][1])\n",
    "    print('COMBOS',len(setsInit[0]),len(setsInit[1]))\n",
    "setsAcrossCombosE = [(set(setDict[c][0]),set(setDict[c][1])) for c in combos]\n",
    "setsInitAll = setsAcrossCombosE[0]\n",
    "for c in range(len(combos)):\n",
    "    setsInitAll = (setsInitAll[0]|setsAcrossCombosE[c][0],setsInitAll[1]|setsAcrossCombosE[c][1])\n",
    "    print('INCL COMBOS',len(setsInitAll[0]),len(setsInitAll[1]))\n",
    "\n",
    "setDict_b = {}\n",
    "rankDict_b = {}\n",
    "for r in ['PP_30','PP_6']:\n",
    "    setDict_b[r] = ([g for g in resDict[r] if resDict[r][g][1]<0.05 and resDict[r][g][0]<0],[g for g in resDict[r] if resDict[r][g][1]<0.05 and resDict[r][g][0]>0])\n",
    "    rankDict_b[r] = {g:resDict[r][g][0] for g in resDict[r]}\n",
    "    print(r,len(setDict_b[r][0]),len(setDict_b[r][1]))\n",
    "rankDict.update(rankDict_b)\n",
    "\n",
    "\n",
    "setDict_bexcl = {}\n",
    "for r in setDict_b:\n",
    "    setDict_bexcl[r] = ([g for g in setDict_b[r][0] if len([_ for r in setDict_b if g in setDict_b[r][0]])==1],[g for g in setDict_b[r][1] if len([_ for r in setDict_b if g in setDict_b[r][1]])==1])\n",
    "    print(r,len(setDict_bexcl[r][0]),len(setDict_bexcl[r][1]))\n",
    "    \n",
    "gmtfile = wDir+'{}.gmt'.format(tid)\n",
    "with open(gmtfile,'w') as o:\n",
    "    for c in cc:\n",
    "        if len(setDict[c][0]):\n",
    "            o.write(\"{}_DN\\tna\\t{}\\n\".format(c,'\\t'.join(setDict[c][0])))\n",
    "        if len(setDict[c][1]):\n",
    "            o.write(\"{}_UP\\tna\\t{}\\n\".format(c,'\\t'.join(setDict[c][1])))\n",
    "    for c in ['PP_30','PP_6']:\n",
    "        if len(setDict_b[c][0]):\n",
    "            o.write(\"{}_DN\\tna\\t{}\\n\".format(c,'\\t'.join(setDict_b[c][0])))\n",
    "        if len(setDict_b[c][1]):\n",
    "            o.write(\"{}_UP\\tna\\t{}\\n\".format(c,'\\t'.join(setDict_b[c][1])))\n",
    "\n",
    "gmtfile = wDir+'{}_combos.gmt'.format(tid)\n",
    "with open(gmtfile,'w') as o:\n",
    "    for c in combos:\n",
    "        if len(setDict[c][0]):\n",
    "            o.write(\"{}_DN\\tna\\t{}\\n\".format(c,'\\t'.join(setDict[c][0])))\n",
    "        if len(setDict[c][1]):\n",
    "            o.write(\"{}_UP\\tna\\t{}\\n\".format(c,'\\t'.join(setDict[c][1])))        \n",
    "    o.write(\"COMBO_DN\\tna\\t{}\\n\".format('\\t'.join(setsInitAll[0])))      \n",
    "    o.write(\"COMBO_UP\\tna\\t{}\\n\".format('\\t'.join(setsInitAll[1])))      \n",
    "            \n",
    "gmtfile = wDir+'{}_excl.gmt'.format(tid)\n",
    "with open(gmtfile,'w') as o:\n",
    "    for c in cc:\n",
    "        if len(setDict_excl[c][0]):\n",
    "            o.write(\"{}excl_DN\\tna\\t{}\\n\".format(c,'\\t'.join(setDict_excl[c][0])))\n",
    "        if len(setDict_excl[c][1]):\n",
    "            o.write(\"{}excl_UP\\tna\\t{}\\n\".format(c,'\\t'.join(setDict_excl[c][1])))\n",
    "    for c in ['PP_30','PP_6']:\n",
    "        if len(setDict_bexcl[c][0]):\n",
    "            o.write(\"{}excl_DN\\tna\\t{}\\n\".format(c,'\\t'.join(setDict_bexcl[c][0])))\n",
    "        if len(setDict_bexcl[c][1]):\n",
    "            o.write(\"{}excl_UP\\tna\\t{}\\n\".format(c,'\\t'.join(setDict_bexcl[c][1])))\n",
    "\n",
    "for r in rankDict:\n",
    "    with open(wDir+\"{}.rnk\".format(r),'w') as o:\n",
    "        for g in [g for g in ggs if g in rankDict[r]]:\n",
    "            o.write('{}\\t{:.2f}\\n'.format(g,round(rankDict[r][g],2)+0))\n",
    "            \n",
    "cmd = \"batchGSEA.py {} {} {}\".format(wDir,gsdir,gspath)\n",
    "print(cmd)\n",
    "os.system(cmd)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ef335e5",
   "metadata": {},
   "outputs": [],
   "source": [
    "pp = PdfPages(\"F3_sigGenesBar.pdf\")\n",
    "rsd = [r for r in resDict if r not in ('PP_30','PP_6')]\n",
    "plt.bar(range(len(rsd)),[len([g for g in resDict[r] if resDict[r][g][-1]<0.05 and resDict[r][g][0]<0]) for r in rsd],color='blue',label='down')\n",
    "plt.bar(range(len(rsd)),[len([g for g in resDict[r] if resDict[r][g][-1]<0.05 and resDict[r][g][0]>0]) for r in rsd],color='red',label='up',bottom=[len([g for g in resDict[r] if resDict[r][g][-1]<0.05 and resDict[r][g][0]<0]) for r in rsd])\n",
    "plt.xticks(range(len(rsd)),rsd,rotation='vertical')\n",
    "plt.legend()\n",
    "plt.title(\"Significant genes by condition\")\n",
    "plt.ylabel(\"Number of significant changes in RA (FDR<0.05)\")\n",
    "plt.xlabel(\"Drug tested\")\n",
    "pp.savefig()\n",
    "#plt.xlim(-0.5,7.5)\n",
    "plt.show()\n",
    "                                                                                                                                       \n",
    "pp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4521435d",
   "metadata": {},
   "outputs": [],
   "source": [
    "#TOP gene tests\n",
    "for r in resDict:\n",
    "    print(r,stats.wilcoxon([resDict[r][g][0] for g in resDict[r] if resDict[r][g][-1]<0.05 and g in classicTOP],correction=True,alternative='less'))\n",
    "    print(r,stats.wilcoxon([resDict[r][g][0] for g in resDict[r] if g in classicTOP],correction=True,alternative='less'))\n",
    "    print(r,stats.mannwhitneyu([resDict[r][g][0] for g in resDict[r] if g in classicTOP],[resDict[r][g][0] for g in resDict[r] if g not in classicTOP],alternative='less'))\n",
    "    print(r,stats.mannwhitneyu([resDict[r][g][0] for g in resDict[r] if g in classicTOP and resDict[r][g][-1]<0.05],[resDict[r][g][0] for g in resDict[r] if g not in classicTOP and resDict[r][g][-1]<0.05],alternative='less'))\n",
    "    \n",
    "    s_T = len([g for g in resDict[r] if resDict[r][g][-1]<0.05 and resDict[r][g][0]<0 and g in classicTOP])\n",
    "    s_nT =len([g for g in resDict[r] if resDict[r][g][-1]<0.05 and resDict[r][g][0]<0  and g not in classicTOP])\n",
    "    ns_T = len([g for g in resDict[r] if resDict[r][g][-1]>=0.05 and resDict[r][g][0]<0  and g in classicTOP])\n",
    "    ns_nT = len([g for g in resDict[r] if resDict[r][g][-1]>=0.5 and resDict[r][g][0]<0  and g not in classicTOP])\n",
    "    print(s_T,s_nT,ns_T,ns_nT)\n",
    "    print(stats.fisher_exact([[s_T,ns_T],[s_nT,ns_nT]]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "2f8544fd",
   "metadata": {},
   "outputs": [],
   "source": [
    "##GENERATE PCA ON ERCC-FREE MATRICES (w/o normalization wrt ERCC)\n",
    "#PC1 correlation vs effect magnitude/significance\n",
    "#TODO: make sure PD022024 PCAs use appropriate (1) outlier lists and (2) underlying genes (i.e. significant in DESeq2)\n",
    "ccmap = plt.get_cmap('Set1')\n",
    "#genes = [g for g in rDict['PN024_ribo'][0].keys() if g not in badG]\n",
    "#erccs = [g for g in genes if 'ERCC-' in g]\n",
    "ddir = wDir\n",
    "pfx = 'riboPLATE.final'\n",
    "for mod in ['ctint','tbasic']:\n",
    "    for norm in ['vst','norm']:\n",
    "        try:\n",
    "                fl = ddir+'{p}.{md}.{norm}.txt'.format(p=pfx,md=mod,norm=norm)\n",
    "                allcts = readMatrix(fl)\n",
    "                ctDict = {s:{g:float(v) for (g,v) in allcts[s].items()} for s in allcts}\n",
    "                riboD = {s.split('_')[1]:v for (s,v) in ctDict.items() if 'ribo' in s}                    \n",
    "                rnaD = {s.split('_')[1]:v for (s,v) in ctDict.items() if 'rna' in s}\n",
    "                sampsA = sorted(list(set([e for e in riboD.keys()])|set([e for e in rnaD.keys()])))\n",
    "                print(len(sampsA))\n",
    "                teD = {k:{v:float(riboD[k][v])-float(rnaD[k][v]) for v in riboD[k] if v in rnaD[k]} for k in riboD if k in rnaD}                \n",
    "                ggenes = [g for g in teD[sampsA[0]].keys() if g in sigGs[0]|sigGs[1]]\n",
    "                print(\"{} genes\".format(len(ggenes)))\n",
    "                lbl = ' '.join([pfx,norm,mod])\n",
    "                pp = PdfPages(\"{}{}_PCA.pdf\".format(outdir,'.'.join(mod,norm)))\n",
    "                condsA = [maplist[int(i)] for i in sampsA]\n",
    "                groupsA = list(set(condsA))\n",
    "                grpsA = {i:e for (i,e) in enumerate(condsA)}                    \n",
    "                conditionIdx = {c:i for i,c in enumerate(groupsA)}\n",
    "                NUM_COLORS = len(groupsA)\n",
    "                colors = {grp:ccmap(conditionIdx[grp]/NUM_COLORS) for grp in groupsA}\n",
    "                dIdx = ['ribo','rna','RA']\n",
    "                for (idx,typD) in [(0,riboD),(1,rnaD),(2,teD)]:\n",
    "                    tmp = [(s,condsA[i]) for (i,s) in enumerate(sampsA) if s in typD] \n",
    "                    samps = [s[0] for s in tmp]\n",
    "                    conds = [s[1] for s in tmp]\n",
    "                    dat = [[typD[s][g] for g in ggenes] for s in samps]\n",
    "                    plt.figure(figsize=(5,5))\n",
    "                    plt.title(dIdx[idx])\n",
    "                    #Change False to True to put sample annotations on each point in the PCA\n",
    "                    plotPCA(plt.gca(),dat,samps,conds,colors,False)\n",
    "                    pp.savefig()\n",
    "                    plt.show()\n",
    "                ctlS = [k for k in teD if maplist[int(k)]=='CTRL']\n",
    "                meanCTL = {g:np.mean([teD[k][g] for k in teD if maplist[int(k)]=='CTRL']) for g in teD[sampsA[0]].keys()}\n",
    "                lfcTE = {k:{g:teD[k][g]-meanCTL[g] for g in teD[k]} for k in teD if maplist[int(k)]!='CTRL'}\n",
    "                ggenes = [g for g in teD[sampsA[0]].keys()]\n",
    "                print(\"{} genes\".format(len(ggenes)))\n",
    "                plt.figure(figsize=(5,5))\n",
    "                sampsE = sorted(list([s for s in lfcTE.keys()]))\n",
    "                cndsE = [maplist[int(e)] for e in sampsE]\n",
    "                dat = [[lfcTE[s][g] for g in ggenes] for s in sampsE]\n",
    "                plotPCA(plt.gca(),dat,sampsE,cndsE,colors,False)\n",
    "                plt.title(\"log-fold change in RA vs DMSO\")\n",
    "                pp.savefig()\n",
    "                plt.show()\n",
    "                \n",
    "                plt.figure()\n",
    "                handles = [mpatches.Patch(color=colors[cond],label=cond) for cond in colors]\n",
    "                plt.legend(handles=handles,frameon=False)\n",
    "                plt.gca().set_axis_off()\n",
    "                pp.savefig()\n",
    "                plt.show()\n",
    "                pp.close()\n",
    "                \n",
    "                meanPC = {c:(np.mean([res[i][0] for i,s in enumerate(sampsE) if maplist[int(s)]==c]),\n",
    "                             np.mean([res[i][1] for i,s in enumerate(sampsE) if maplist[int(s)]==c]),                              \n",
    "                             np.mean([np.mean([np.abs(lfcTE[s][g]) for s in lfcTE if maplist[int(s)]==c]) for g in sigGs[0]|sigGs[1] if g in resDict[''.join(c.split('+'))]]),\n",
    "                             len([g for g in resDict[''.join(c.split('+'))] if resDict[''.join(c.split('+'))][g][1]<0.05]),\n",
    "                            )\n",
    "                          for c in set(maplist) if c!='CTRL'}\n",
    "                indPC = {s:(res[i][0],res[i][1],\n",
    "                            np.mean([np.abs(lfcTE[s][g]) for g in sigGs[0]|sigGs[1]])\n",
    "                           ) for (i,s) in enumerate(sampsE)}\n",
    "                plt.scatter([meanPC[c][0] for c in meanPC],[meanPC[c][3] for c in meanPC],c=[colors[c] for c in meanPC])\n",
    "                plt.xlabel(\"PC1\")\n",
    "                plt.ylabel(\"Number of significant effects (across condition)\")\n",
    "                plt.show()\n",
    "                print(stats.spearmanr([meanPC[c][0] for c in meanPC],[meanPC[c][3] for c in meanPC]))\n",
    "                plt.scatter([indPC[s][0] for s in indPC],[indPC[s][2] for s in indPC],c=[colors[maplist[int(s)]] for s in indPC])\n",
    "                plt.xlabel(\"PC1\")\n",
    "                plt.ylabel(\"mean absolute effect (individual samples)\")\n",
    "                plt.show()                \n",
    "                print(stats.spearmanr([indPC[s][0] for s in indPC],[indPC[s][2] for s in indPC]))                       \n",
    "\n",
    "            except OSError:\n",
    "                print(fl)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "c4d08b7b",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Depletion scatterplots -- use tbasic vst for sake of comparison (or norm, since it's closely related to depletion histograms w norm counts)\n",
    "for mod in ['tbasic','ctint']:\n",
    "    for norm in ['norm','vst']:\n",
    "        for pfx in ['FL','riboPLATE','riboPLATE_ERCC']:#[::-1]:                \n",
    "            try:\n",
    "                fl = wDir+'{p}.final.{md}.{norm}.txt'.format(p=pfx,md=mod,norm=norm)\n",
    "                allcts = readMatrix(fl)\n",
    "            except Exception:\n",
    "                print (mod,norm,pfx,\"NOT FOUND\")\n",
    "                continue\n",
    "                \n",
    "            ctDict = {s:{g:float(v) for (g,v) in allcts[s].items()} for s in allcts}\n",
    "            if 'FL' == pfx:\n",
    "                riboD = {'_'.join(s.split('_')[:-1]):v for (s,v) in ctDict.items() if 'RIBO' in s}\n",
    "                rnaD = {'_'.join(s.split('_')[:-1]):v for (s,v) in ctDict.items() if 'RNA' in s}\n",
    "            else:\n",
    "                riboD = {s.split('_')[1]:v for (s,v) in ctDict.items() if 'ribo' in s}                    \n",
    "                rnaD = {s.split('_')[1]:v for (s,v) in ctDict.items() if 'rna' in s}\n",
    "            sampsA = sorted(list(set([e for e in riboD.keys()])|set([e for e in rnaD.keys()])))\n",
    "            print(len(sampsA))\n",
    "            teD = {k:{v:float(riboD[k][v])-float(rnaD[k][v]) for v in riboD[k] if v in rnaD[k]} for k in riboD if k in rnaD}                \n",
    "            ggenes = [g for g in teD[sampsA[0]].keys()]# if g in sigG[0]+sigG[1]]                \n",
    "            #ncRNA scatterplots\n",
    "            lbl='.'.join((pfx,norm,mod))\n",
    "            pp = PdfPages(\"{}{}_depletion_scatter.pdf\".format(outdir,lbl))                    \n",
    "            dict1 = rnaD\n",
    "            dict2 = teD\n",
    "            samps = [s for s in dict1 if s in dict2]\n",
    "            print(len(samps))\n",
    "            plt.figure(figsize=(5,5))\n",
    "            plt.title('_'.join([pfx,mod,norm]))\n",
    "            if pfx=='riboPLATE_ERCC':\n",
    "                smpE = [str(e) for e in range(96)[::2] if str(e) in sampsA]\n",
    "            else:\n",
    "                smpE = samps\n",
    "            plt.scatter([np.mean([rnaD[s][g] for s in smpE]) for g in ggenes if g not in ncGenes and 'ERCC-' not in g],[np.mean([teD[s][g] for s in smpE]) for g in ggenes if g not in ncGenes and 'ERCC-' not in g],label='mRNA',color='blue',alpha=0.5,s=20,edgecolors='none')\n",
    "            plt.scatter([np.mean([rnaD[s][g] for s in smpE]) for g in ggenes if g in ncGenes],[np.mean([teD[s][g] for s in smpE]) for g in ggenes if g in ncGenes],label='ncRNA',color='red',alpha=0.5,s=20,edgecolors='none')\n",
    "            plt.legend()\n",
    "            print(stats.wilcoxon([np.mean([teD[s][g]/rnaD[s][g] for g in ggenes if g in ncGenes and 'ERCC-' not in g]) for s in smpE], [np.mean([teD[s][g]/rnaD[s][g] for g in ggenes if g not in ncGenes and 'ERCC-' not in g]) for s in smpE],correction=True,alternative='greater'))\n",
    "            plt.xlabel('RNA')#l.split('V')[0])\n",
    "            plt.ylabel('RA')\n",
    "            pp.savefig()\n",
    "            plt.show()\n",
    "            if pfx == 'PN024':\n",
    "                plt.figure(figsize=(5,5))\n",
    "                plt.title('_'.join([pfx,mod,norm]))\n",
    "                ercG = [g for g in ggenes if 'ERCC-' in g]\n",
    "                nrcG = [g for g in ggenes if 'ERCC-' not in g and g not in ncGenes]\n",
    "                xs = [np.mean([rnaD[s][g] for s in smpE]) for g in nrcG]\n",
    "                ys = [np.mean([teD[s][g] for s in smpE]) for g in nrcG]\n",
    "                plt.scatter(xs,ys,label='mRNA',color='blue',alpha=0.5,s=20,edgecolors='none')\n",
    "                xs = [np.mean([rnaD[s][g] for s in smpE]) for g in ercG]\n",
    "                ys = [np.mean([teD[s][g] for s in smpE]) for g in ercG]\n",
    "                plt.scatter(xs,ys,label='ERCC',color='red',alpha=0.5,s=20,edgecolors='none')\n",
    "                plt.xlabel(\"RNA\")\n",
    "                plt.ylabel(\"RA\")\n",
    "                print(stats.wilcoxon([np.mean([teD[s][g] for g in nrcG]) for s in smpE], [np.mean([teD[s][g] for g in ercG]) for s in smpE],correction=True,alternative='greater'))\n",
    "                plt.legend()\n",
    "                pp.savefig()\n",
    "                plt.show()\n",
    "            pp.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "da2fa794",
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in ['PP242','AZD','BKM','MNK1','4EGI1']:#,'PP_6','PP_30']:\n",
    "    pp = PdfPages(\"{}{}_volcano.pdf\".format(outdir,c))\n",
    "    plt.figure(figsize=(6,6))\n",
    "    ymax=10\n",
    "    if c!='4EGI1':\n",
    "        xm = 2.5\n",
    "    else:\n",
    "        xm = 7.9\n",
    "    #In loading resDict, p values of NA converted to 1 --> many changes w/ no significance computed b/c outliers or overdispersion\n",
    "    #Must be filtered for the sake of displaying volcano plot, but what about in ranked lists and other comparisons across conditions?\n",
    "    #If idea is to demonstrate effect sizes w/o regard to significance, leave them in\n",
    "    #If that's not a realistic application of the technology and significance should be restricted, how do things change downstream?\n",
    "    \n",
    "    plt.scatter([resDict[c][g][0] for g in resDict[c] if g not in classicTOP and resDict[c][g][1]!=1.0],[-np.log10(resDict[c][g][1]) if resDict[c][g][1]!=0 and -np.log10(resDict[c][g][1])<ymax else ymax for g in resDict[c] if g not in classicTOP and resDict[c][g][1]!=1.0],alpha=0.7,s=30,label='non-TOP',color='black',edgecolors='none')\n",
    "    #plt.scatter([resDict[c][g][0] for g in genes if g in allTOP],[-np.log10(resDict[c][g][1]+1e-40) for g in genes if g in allTOP],alpha=0.5,s=3,label='TOP',color='red')\n",
    "    plt.scatter([resDict[c][g][0] for g in resDict[c] if g in classicTOP and resDict[c][g][1]!=1.0],[-np.log10(resDict[c][g][1]) if resDict[c][g][1]!=0 and -np.log10(resDict[c][g][1])<ymax else ymax for g in resDict[c] if g in classicTOP and resDict[c][g][1]!=1.0],alpha=0.85,s=30,label='TOP',color='green',edgecolors='none')#n([resDict[c][g][1] for g in resDict[c] if resDict[c][g][1]])/10\n",
    "    plt.axhline(-np.log10(0.05),linestyle='dashed',color='teal',alpha=0.5)\n",
    "    plt.title(c)\n",
    "    #plt.xlim(-5,5)\n",
    "    plt.ylim(0,ymax+0.2)\n",
    "    plt.xlim(-xm,xm)\n",
    "    plt.axvline(1,linestyle='dashed',alpha=0.5,color='teal')\n",
    "    plt.axvline(-1,linestyle='dashed',alpha=0.5,color='teal')\n",
    "    plt.title(c)\n",
    "    plt.xlabel(\"Log-Fold Change in RA, vs DMSO\")\n",
    "    plt.ylabel(\"Significance (-log(FDR))\")\n",
    "    plt.legend()\n",
    "    pp.savefig()\n",
    "    plt.show()\n",
    "    print(stats.mannwhitneyu([resDict[c][g][0] for g in resDict[c] if g in classicTOP],[resDict[c][g][0] for g in resDict[c] if g not in classicTOP],alternative='less',use_continuity=True))    \n",
    "    pp.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b5e568f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for c in ['PP_6','PP_30']:\n",
    "    pp = PdfPages(\"{}{}_volcano.pdf\".format(outdir,c))\n",
    "    plt.figure(figsize=(6,6))\n",
    "    ymax=10\n",
    "    #MAKE TOP GENES GREEN, RIBOPLATE TRIANGLES CYAN/MAGENTA, AND BG COLOR BLUE\n",
    "    #In loading resDict, p values of NA converted to 1 --> many changes w/ no significance computed b/c outliers or overdispersion\n",
    "    #Must be filtered for the sake of displaying volcano plot, but what about in ranked lists and other comparisons across conditions?\n",
    "    #If idea is to demonstrate effect sizes w/o regard to significance, leave them in\n",
    "    #If that's not a realistic application of the technology and significance should be restricted, how do things change downstream?\n",
    "    ppTarg = [g for g in resDict['PP242'] if resDict['PP242'][g][-1]<0.05 and resDict['PP242'][g][0]<0],[g for g in resDict['PP242'] if resDict['PP242'][g][-1]<0.05 and resDict['PP242'][g][0]>0]\n",
    "    #scatter background genome -- not TOP, not target of PP242 in riboPLATE\n",
    "    xs = [resDict[c][g][0] for g in resDict[c] if g not in classicTOP and resDict[c][g][1]!=1.0 and g not in ppTarg[0]+ppTarg[1]]\n",
    "    ys = [-np.log10(resDict[c][g][1]) if resDict[c][g][1]!=0 and -np.log10(resDict[c][g][1])<ymax else ymax for g in resDict[c] if g not in classicTOP and resDict[c][g][1]!=1.0 and g not in ppTarg[0]+ppTarg[1]]\n",
    "    plt.scatter(xs,ys,color='black',label='background',s=20,alpha=0.3,edgecolors='none')\n",
    "    xs = [resDict[c][g][0] for g in resDict[c] if g in ppTarg[1]]\n",
    "    ys = [-np.log10(resDict[c][g][1]) if resDict[c][g][1]!=0 and -np.log10(resDict[c][g][1])<ymax else ymax for g in resDict[c] if g in ppTarg[1]]\n",
    "    plt.scatter(xs,ys,color='red',alpha=0.7,s=30,edgecolors='none',label='riboPLATE up')\n",
    "    xs = [resDict[c][g][0] for g in resDict[c] if g in ppTarg[0]]\n",
    "    ys = [-np.log10(resDict[c][g][1]) if resDict[c][g][1]!=0 and -np.log10(resDict[c][g][1])<ymax else ymax for g in resDict[c] if g in ppTarg[0]]\n",
    "    plt.scatter(xs,ys,color='blue',alpha=0.7,s=30,edgecolors='none',label='riboPLATE down')\n",
    "\n",
    "    xs = [resDict[c][g][0] for g in resDict[c] if g in classicTOP and resDict[c][g][1]!=1.0]# and g not in ppTarg[0]+ppTarg[1]]\n",
    "    ys = [-np.log10(resDict[c][g][1]) if resDict[c][g][1]!=0 and -np.log10(resDict[c][g][1])<ymax else ymax for g in resDict[c] if g in classicTOP and resDict[c][g][1]!=1.0]# and resDict[c][g][1]!=1.0 and g not in ppTarg[0]+ppTarg[1]]    \n",
    "    plt.scatter(xs,ys,color='green',label='TOP',alpha=0.5,s=30,edgecolors='None')\n",
    "    \n",
    "    plt.axhline(y=-np.log10(0.05),linestyle='dashed',color='teal',alpha=0.5)\n",
    "    plt.axvline(x=-1,linestyle='dashed',color='teal',alpha=0.5)\n",
    "    plt.axvline(x=1,linestyle='dashed',color='teal',alpha=0.5)\n",
    "    xmax = max([np.abs(resDict[c][g][0]) for g in resDict[c]])\n",
    "    if c=='PP_6':\n",
    "        plt.xlim(-7,7)\n",
    "    elif c=='PP_30':\n",
    "        plt.xlim(-11,11)\n",
    "    plt.ylim(0,ymax+0.2)\n",
    "    plt.title(c)\n",
    "    plt.xlabel(\"log-fold change in TE, vs DMSO\")\n",
    "    plt.ylabel(\"Significance (-log(FDR))\")\n",
    "    plt.legend()\n",
    "    pp.savefig()\n",
    "    plt.show()\n",
    "    print(stats.mannwhitneyu([resDict[c][g][0] for g in resDict[c] if g in classicTOP],[resDict[c][g][0] for g in resDict[c] if g not in classicTOP],use_continuity=True,alternative='less'))    \n",
    "    print(stats.mannwhitneyu([resDict[c][g][0] for g in resDict[c] if g in ppTarg[0]],[resDict[c][g][0] for g in resDict[c] if g in ppTarg[1]],use_continuity=True,alternative='less'))    \n",
    "\n",
    "    pp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "17df7149",
   "metadata": {},
   "outputs": [],
   "source": [
    "#GSEA 1: heatmap of plate-riboseq effect correlation\n",
    "gsdir = gsDir+'out/'\n",
    "gsDict = {}\n",
    "for rnk in ['PP242','MNK1','AZD','BKM','4EGI1','PP_6','PP_30']:#['PP_6','PP_30']:\n",
    "    dfl = [e for e in os.listdir(gsdir) if rnk in e.split('.')[0] and 'riboPLATE' in e.split('.')[0] and 'combos' not in e and 'excl' in e]\n",
    "    dfl = dfl[0]\n",
    "    print(dfl)\n",
    "    dpos = [e for e in os.listdir(gsdir+dfl) if 'gsea_report_for_na_pos' in e and '.tsv' in e]\n",
    "    dneg = [e for e in os.listdir(gsdir+dfl) if 'gsea_report_for_na_neg' in e and '.tsv' in e]\n",
    "    if len(dpos)>1 or len(dneg)>1:\n",
    "        print(\"PROBLEM\")\n",
    "        print(dpos)\n",
    "        print(dneg)\n",
    "    dpos = dpos[0]\n",
    "    dneg = dneg[0]\n",
    "    with open(gsdir+dfl+'/'+dpos,'r') as fp, open(gsdir+dfl+'/'+dneg,'r') as fn:\n",
    "        fp.readline()\n",
    "        fn.readline()\n",
    "        gsDict[rnk] = {e.split('\\t')[0]:(e.split('\\t')[5:9]) for e in fp.readlines()}\n",
    "        gsDict[rnk].update({e.split('\\t')[0]:(e.split('\\t')[5:9]) for e in fn.readlines()})\n",
    "sets = sorted([s for s in gsDict[rnk] if '_30' not in s and '_6' not in s])\n",
    "sets = [s for s in sets if '_DN' in s][::-1]+[s for s in sets if '_UP' in s][::-1]\n",
    "rnks = ['PP_6','PP_30']#,'4EGI1','MNK1','AZD','BKM','PP242']\n",
    "#dat = [[float(gsDict[r][s][0]) for r in rnks if '_30' in r or '_6'  in r] for s in sets if '_30' not in s and '_6' not in s]# for r in rnks]\n",
    "dat = [[float(gsDict[r][s][0]) for r in rnks] for s in sets]# for r in rnks]\n",
    "nsig = [[\"*\" if float(gsDict[r][s][-1])<0.05 else \"\" for r in rnks] for s in sets]\n",
    "\n",
    "#dat = [[plateRes[c][g][0] for c in condes] for g in gset]\n",
    "#rowcol = ['g' if g in allTOP else 'black' for g in gset]\n",
    "metr = 'correlation'\n",
    "#for metr in ['correlation','euclidean','seuclidean']:\n",
    "pp = PdfPages(outdir+\"riboPLATErpfGSEA.pdf\")\n",
    "plt.figure(figsize=(4,6))\n",
    "sw = sns.heatmap(dat,xticklabels=rnks,annot=nsig,fmt='',yticklabels=sets,cmap='bwr',center=0,cbar_kws={'label':'GSEA Normalized Enrichment Score (NES)'})#,square=True)#,figsize=(5,10))\n",
    "#    plt.title(\"{} distance\".format(metr))\n",
    "sw.set_xlabel(\"Drug Tested (log-fold change in TE vs DMSO, ribosome profiling/RNA-seq)\")\n",
    "sw.set_ylabel(\"Gene Sets of Drug Targets (FDR<0.05, riboPLATE/PLATE-seq RA)\")\n",
    "pp.savefig()\n",
    "plt.show()\n",
    "pp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "649d724d",
   "metadata": {},
   "outputs": [],
   "source": [
    "ymax = 10\n",
    "colors = it.cycle(('blue','red','cyan','magenta'))\n",
    "for combo in [('PP242','BKM'),('PP242','MNK1'),('MNK1','BKM')]:\n",
    "    cset = ''.join(combo)\n",
    "    pp = PdfPages(outdir+\"{}_combo_volcano.pdf\".format(cset))\n",
    "    plt.figure(figsize=(6,6))\n",
    "    gs = set()\n",
    "    for c in combo:\n",
    "        gs_dn = [g for g in resDict[c] if resDict[c][g][-1]<0.05 and resDict[c][g][0]<0]#setDict_b[cset][0]\n",
    "        gs_up = [g for g in resDict[c] if resDict[c][g][-1]<0.05 and resDict[c][g][0]>0]#setDict_b[cset][1]\n",
    "        gs |= set(gs_dn)\n",
    "        gs |= set(gs_up)\n",
    "    plt.scatter([resDict[cset][g][0] for g in resDict[cset] if g not in gs and resDict[cset][g][-1]!=1],[-np.log10(resDict[cset][g][-1]) if resDict[cset][g][-1]!=0 and -np.log10(resDict[cset][g][-1])<ymax else ymax for g in resDict[cset] if g not in gs and resDict[cset][g][-1]!=1],color='black',alpha=0.5,label='bg',edgecolors='none')\n",
    "    for c in combo:\n",
    "        gs_dn = [g for g in resDict[c] if resDict[c][g][-1]<0.05 and resDict[c][g][0]<0]#setDict_b[cset][0]\n",
    "        gs_up = [g for g in resDict[c] if resDict[c][g][-1]<0.05 and resDict[c][g][0]>0]#setDict_b[cset][1]    \n",
    "        plt.scatter([resDict[cset][g][0] for g in resDict[cset] if g in gs_dn and resDict[cset][g][-1]!=1],[-np.log10(resDict[cset][g][-1]) if resDict[cset][g][-1]!=0 and -np.log10(resDict[cset][g][-1])<ymax else ymax for g in resDict[cset] if g in gs_dn and resDict[cset][g][-1]!=1],color=next(colors),alpha=0.3,edgecolors='none',label='{}_dn'.format(c))\n",
    "        plt.scatter([resDict[cset][g][0] for g in resDict[cset] if g in gs_up and resDict[cset][g][-1]!=1],[-np.log10(resDict[cset][g][-1]) if resDict[cset][g][-1]!=0 and -np.log10(resDict[cset][g][-1])<ymax else ymax for g in resDict[cset] if g in gs_up and resDict[cset][g][-1]!=1],color=next(colors),alpha=0.3,edgecolors='none',label='{}_up'.format(c))\n",
    "    plt.title(\"{}\".format(cset))\n",
    "    plt.ylim(0,ymax+0.2)\n",
    "    plt.xlim(-5,5)\n",
    "    plt.axhline(-np.log10(0.05),linestyle='dashed')\n",
    "    plt.axvline(-1,alpha=0.5)\n",
    "    plt.axvline(1,alpha=0.5)\n",
    "    plt.legend()\n",
    "    pp.savefig()\n",
    "    plt.show()\n",
    "    pp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "528b8092",
   "metadata": {},
   "outputs": [],
   "source": [
    "cc = ['PP242','MNK1','4EGI1','BKM','AZD']#'MNK1BKM','PP242BKM','PP242MNK1']\n",
    "combos = [('MNK1','BKM'),('PP242','BKM'),('PP242','MNK1')]\n",
    "#Easier plot to digest: for each drug combo, plot effect size in singular vs effect size in combo for effects of both\n",
    "#individual drugs\n",
    "ccmap = plt.get_cmap('Set1')\n",
    "grpsA = {i:e for (i,e) in enumerate(set(maplist))}                    \n",
    "conditionIdx = {c:i for i,c in enumerate(set(maplist))}\n",
    "NUM_COLORS = len(grpsA)\n",
    "colors = {c:ccmap(conditionIdx[c]/NUM_COLORS) for c in conditionIdx}\n",
    "for comb in combos:\n",
    "    c = ''.join(comb)\n",
    "    pp = PdfPages(outdir+\"{}_combo_attenuation.pdf\".format(c))    \n",
    "    #gs: dict of gene:(effect,condition) for max effect seen in either individual condition\n",
    "    gs = {}\n",
    "    for ci in comb:\n",
    "        for g in [g for g in resDict[ci] if resDict[ci][g][-1]<0.05]:\n",
    "            try:\n",
    "                if np.abs(resDict[ci][g][0])>np.abs(gs[g][0]):\n",
    "                    gs[g] = (resDict[ci][g][0],ci)\n",
    "                #gs[g] = np.max(np.abs(gs[g]),np.abs(resDict[ci][g][0]))\n",
    "            except Exception:\n",
    "                gs[g] = (resDict[ci][g][0],ci)\n",
    "        #gs = {g:resDict[ci][g][0] for g in resDict[ci] if resDict[ci][g][-1]<0.05}\n",
    "        #for g in gs:\n",
    "        #    resDict[ci][g][-1]\n",
    "        #gs  |= set([g for g in resDict[ci] if resDict[ci][g][-1]<0.05])\n",
    "    #domX: set of genes significant in combination treatment\n",
    "    domX = [g for g in resDict[c] if resDict[c][g][-1]<0.05]\n",
    "    #ranX: set of genes significant in either individual treatment\n",
    "    ranX = [g for g in gs]\n",
    "    #xs = [resDict[c][g][0] for g in ranX]\n",
    "    plt.figure(figsize=(5,5))\n",
    "\n",
    "    #xs: effect size in individual drug of max effect\n",
    "    xs = [gs[g][0] for g in ranX if g in resDict[c]]\n",
    "    #ys: effect size in combination treatment\n",
    "    ys = [resDict[c][g][0] for g in ranX if g in resDict[c]]\n",
    "    #colors: condition in which max absolute effect is seen for target\n",
    "    cs = [colors[gs[g][1]] for g in ranX if g in resDict[c]]\n",
    "    plt.scatter(xs,ys,c=cs,alpha=0.75,edgecolors='none')\n",
    "    plt.xlim(-3,3)\n",
    "    plt.ylim(-3,3)\n",
    "    #plt.ylim(-3.5,3.5)\n",
    "    plt.axhline(0,linestyle='dashed',color='black')\n",
    "    handles = [mpatches.Patch(color=colors[cond],label=cond) for cond in comb]\n",
    "    plt.legend(handles=handles)\n",
    "    #plt.axvline(0)\n",
    "    plt.plot(range(-4,4),range(-4,4),linestyle='dashed',color='black')\n",
    "    plt.title(c)\n",
    "    plt.xlabel(\"Maximum effect observed in singular drug {}\".format(comb))\n",
    "    plt.ylabel(\"Observed effect in combination treatment\")\n",
    "    pp.savefig()\n",
    "    plt.show()\n",
    "    pp.close()\n",
    "    #stdev of single drugs' upregulated/downregulated targets in combination treatment effect \n",
    "    print(\"Stdev up \",np.std([resDict[c][g][0] for g in gs if g in resDict[c] and resDict[gs[g][1]][g][0]>0]))\n",
    "    print(\"Stdev dn \",np.std([resDict[c][g][0] for g in gs if g in resDict[c] and resDict[gs[g][1]][g][0]<0]))\n",
    "\n",
    "    print(len([g for g in gs if np.sign(resDict[c][g][0])!=np.sign(resDict[gs[g][1]][g][0])])/len(gs),\n",
    "         len([g for g in gs if np.sign(resDict[c][g][0])!=np.sign(resDict[gs[g][1]][g][0]) and resDict[gs[g][1]][g][0]>0])/len([g for g in gs if resDict[gs[g][1]][g][0]>0]),\n",
    "         len([g for g in gs if np.sign(resDict[c][g][0])!=np.sign(resDict[gs[g][1]][g][0]) and resDict[gs[g][1]][g][0]<0])/len([g for g in gs if resDict[gs[g][1]][g][0]<0]),\n",
    "         )\n",
    "    print(\"percent attenuated overall \",len([g for g in gs if np.abs(resDict[c][g][0])<np.abs(resDict[gs[g][1]][g][0])])/len(gs),\n",
    "          \"percent attenuated up \",len([g for g in gs if np.abs(resDict[c][g][0])<np.abs(resDict[gs[g][1]][g][0]) and resDict[gs[g][1]][g][0]>0])/len([g for g in gs if resDict[gs[g][1]][g][0]>0]),\n",
    "          \"percent attenuated down \",len([g for g in gs if np.abs(resDict[c][g][0])<np.abs(resDict[gs[g][1]][g][0]) and resDict[gs[g][1]][g][0]<0])/len([g for g in gs if resDict[gs[g][1]][g][0]<0])\n",
    "         )\n",
    "\n",
    "    print(\"mean attenuation down \", np.mean([resDict[gs[g][1]][g][0]-resDict[c][g][0] for g in gs if resDict[gs[g][1]][g][0]<0 and resDict[c][g][0]>resDict[gs[g][1]][g][0]]),#/len([g for g in gs if resDict[gs[g][1]][g][0]<0]),\n",
    "          \"mean attenuation up \",np.mean([resDict[gs[g][1]][g][0]-resDict[c][g][0]  for g in gs if resDict[gs[g][1]][g][0]>0 and resDict[c][g][0]<resDict[gs[g][1]][g][0]]),#/len([g for g in gs if resDict[gs[g][1]][g][0]>0]),\n",
    "         )       \n",
    "#stdev of observed effect in combination treatment for upregulated vs downregulated?\n",
    "#Proportion that change sign\n",
    "#Mean reduction up/dn for each combo"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "12829440",
   "metadata": {},
   "outputs": [],
   "source": [
    "topG = (set([nmDict[t] for t in tpcm if t in nmDict and nmDict[t] not in classicTOP]),[g for g in classicTOP])\n",
    "\n",
    "pp = PdfPages(outdir+\"TOPcandidateRA.pdf\")\n",
    "gIdx = ['mouse_homologs','classic_TOP']\n",
    "cIdx = ['blue','green']\n",
    "plt.figure(figsize=(6,4))\n",
    "#dat = [[resDict[c][g][0] for g in resDict[c] if g in resDict[c] and resDict[c][g][-1]<0.05] for c in ['PP242','AZD','BKM','MNK1','4EGI1']]#,'PP_6','PP_30']]#[[resDict[c][g][0] for c in ['PP242','AZD','BKM','MNK1','4EGI1'] if g in resDict[c] and resDict[c][g][-1]<0.05] for g in gset]\n",
    "#sns.stripplot(data=dat,color='black',size=4)\n",
    "rnks = ['PP242','AZD','BKM','MNK1','4EGI1']\n",
    "for i,gset in enumerate(topG):\n",
    "    print(len(gset))\n",
    "    for c in rnks:#,'PP_6','PP_30']:\n",
    "        print(c,len([g for g in gset if g in resDict[c] and resDict[c][g][-1]<0.05]))\n",
    "    dat = [[resDict[c][g][0] for g in gset if g in resDict[c] and resDict[c][g][-1]<0.05] for c in rnks]#[[resDict[c][g][0] for c in ['PP242','AZD','BKM','MNK1','4EGI1'] if g in resDict[c] and resDict[c][g][-1]<0.05] for g in gset]\n",
    "    sns.stripplot(data=dat,color=cIdx[i],alpha=0.5,size=6,edgecolors='none')\n",
    "plt.xticks(range(len(rnks)),rnks)\n",
    "handles = [mpatches.Patch(color=cIdx[i],label=gIdx[i]) for i in range(2)]\n",
    "#plt.ylim(-2.7,0)\n",
    "plt.legend(handles=handles,loc='lower left')\n",
    "plt.ylabel(\"Log-Fold Change in RA, vs DMSO (p<0.05)\")\n",
    "plt.xlabel(\"Drug Tested\")\n",
    "plt.title(\"Changes in Ribosome Association across TOP Genes & Candidates by Drug\")\n",
    "pp.savefig()\n",
    "plt.show()\n",
    "pp.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7d11c52e",
   "metadata": {},
   "outputs": [],
   "source": [
    "#ERCC depletion histograms for riboPLATEseq experiment\n",
    "ddir = '/home/ubuntu/datadrive/riboPLATEseq/DESeq2/'\n",
    "mod = 'tbasic'\n",
    "norm = 'normB'\n",
    "pfx = 'riboPLATE_ERCC'\n",
    "fl = wDir+'{p}.final.{md}.{norm}.txt'.format(p=pfx,md=mod,norm=norm)\n",
    "allcts = readMatrix(fl)\n",
    "ctDict = {s:{g:float(v) for (g,v) in allcts[s].items()} for s in allcts}\n",
    "riboD = {int(s.split('_')[1]):v for (s,v) in ctDict.items() if 'ribo' in s}                    \n",
    "rnaD = {int(s.split('_')[1]):v for (s,v) in ctDict.items() if 'rna' in s}\n",
    "sampsA = sorted(list(set([e for e in riboD.keys()])|set([e for e in rnaD.keys()])))\n",
    "print(len(sampsA))\n",
    "teD = {k:{v:((float(riboD[k][v])+1)/(float(rnaD[k][v])+1)) for v in riboD[k] if v in rnaD[k]} for k in riboD if k in rnaD}\n",
    "sampsT = [k for k in teD.keys()]\n",
    "ggenes = [g for g in teD[sampsT[0]].keys()]# if g in sigG[0]+sigG[1]]               \n",
    "ercs = [g for g in ggenes if 'ERCC-' in g]\n",
    "print(len(ercs))\n",
    "ss = [i for i in range(96)[::2]]\n",
    "\n",
    "pp = PdfPages(outdir+'ERCCdepletion_sum.pdf')\n",
    "_,bns,_ = plt.hist([((sum([riboD[i][g] for g in ggenes if g not in ercs and rnaD[i][g]!=0]))/(sum([rnaD[i][g] for g in ggenes if g not in ercs and rnaD[i][g]!=0]))) for i in ss],bins=25,range=(0,2.0),alpha=0.5,color='blue',label='genome')\n",
    "plt.hist([((sum([riboD[i][g] for g in ercs if g in riboD[i] and rnaD[i][g]!=0])/(sum([rnaD[i][g] for g in ercs if g in rnaD[i] and rnaD[i][g]!=0])))) for i in ss],bins=bns,alpha=0.5,color='red',label='spike-in')\n",
    "plt.title(\"RA ratios in ERCC spike-in vs genes, summed\")\n",
    "plt.ylabel(\"Wells\")\n",
    "plt.xlabel(\"RA (sum(riboPLATE)/sum(RNA PLATE))\")\n",
    "plt.legend()\n",
    "pp.savefig()\n",
    "plt.show()\n",
    "pp.close()\n",
    "\n",
    "#nmG = [g for g in ggenes if g in nmDict.values()]\n",
    "ercG = [g for g in ggenes if g in ercs]\n",
    "nrcG = [g for g in ggenes if g not in ercs]\n",
    "#Sum RA for non-ERCCs\n",
    "print(np.mean([np.log2((sum([riboD[i][g] for g in ggenes if g not in ercs and rnaD[i][g]!=0]))/(sum([rnaD[i][g] for g in ggenes if g not in ercs and rnaD[i][g]!=0]))) for i in ss]))\n",
    "#Sum RA for ERCCs\n",
    "print(np.mean([np.log2((sum([riboD[i][g] for g in ggenes if g in ercs and rnaD[i][g]!=0]))/(sum([rnaD[i][g] for g in ggenes if g in ercs and rnaD[i][g]!=0]))) for i in ss]))\n",
    "print(stats.mannwhitneyu([np.log2((sum([riboD[i][g] for g in ggenes if g not in ercs and rnaD[i][g]!=0]))/(sum([rnaD[i][g] for g in ggenes if g not in ercs and rnaD[i][g]!=0]))) for i in ss],\n",
    "                         [np.log2((sum([riboD[i][g] for g in ggenes if g in ercs and rnaD[i][g]!=0]))/(sum([rnaD[i][g] for g in ggenes if g in ercs and rnaD[i][g]!=0]))) for i in ss],alternative='greater'))\n",
    "print(stats.wilcoxon([np.log2((sum([riboD[i][g] for g in nrcG]))/(sum([rnaD[i][g] for g in nrcG]))) for i in ss],\n",
    "                         [np.log2((sum([riboD[i][g] for g in ercG]))/(sum([rnaD[i][g] for g in ercG]))) for i in ss],correction=True,alternative='greater'))\n",
    "print(np.mean([np.log2(np.mean([teD[i][g] for g in ggenes if g not in ercs and rnaD[i][g]!=0])) for i in ss]))\n",
    "print(np.mean([np.log2(np.mean([teD[i][g] for g in ggenes if g in ercs and rnaD[i][g]!=0])) for i in ss]))\n",
    "print(stats.mannwhitneyu([np.log2(np.mean([teD[i][g] for g in nrcG])) for i in ss],[np.log2(np.mean([teD[i][g] for g in ercG])) for i in ss],alternative='greater'))\n",
    "print(stats.wilcoxon([np.log2(np.mean([teD[i][g] for g in nrcG])) for i in ss],[np.log2(np.mean([teD[i][g] for g in ercG])) for i in ss],correction=True,alternative='greater'))\n",
    "\n",
    "\n",
    "##DIFFERENCE IN RA -- ERCC VS WHOLE GENOME--SUM (yields log2 RA depletion ratio)\n",
    "pp = PdfPages(outdir+\"ERCCdepletion_logsum.pdf\")\n",
    "_,bns,_ = plt.hist([np.log2((sum([riboD[i][g] for g in ercG if rnaD[i][g]!=0])/(sum([rnaD[i][g] for g in ercG if rnaD[i][g]!=0]))))-np.log2((sum([riboD[i][g] for g in nrcG if rnaD[i][g]!=0]))/(sum([rnaD[i][g] for g in nrcG if rnaD[i][g]!=0]))) for i in ss],bins=25)\n",
    "plt.title(\"ERCC RA depletion ratios, summed\")\n",
    "plt.xlabel(\"log2(RA(ERCC/genome))\")\n",
    "plt.ylabel(\"Wells\")\n",
    "pp.savefig()\n",
    "plt.show()\n",
    "pp.close()\n",
    "#Mean difference in all wells\n",
    "print(np.mean([[np.log2((sum([riboD[i][g] for g in ercG])/(sum([rnaD[i][g] for g in ercG]))))-np.log2((sum([riboD[i][g] for g in nrcG]))/(sum([rnaD[i][g] for g in nrcG]))) for i in ss]]))# for i in ss]]))\n",
    "#Mean difference excluding the two problematic wells\n",
    "print(np.mean([[np.log2((sum([riboD[i][g] for g in ercG])/(sum([rnaD[i][g] for g in ercG]))))-np.log2((sum([riboD[i][g] for g in nrcG]))/(sum([rnaD[i][g] for g in nrcG]))) for i in ss if np.log2(sum([riboD[i][g] for g in ercG])/sum([rnaD[i][g] for g in ercG]))<0]]))# for i in ss]]))\n",
    "print(stats.wilcoxon([np.log2((sum([riboD[i][g] for g in ercG])/(sum([rnaD[i][g] for g in ercG]))))-np.log2((sum([riboD[i][g] for g in nrcG]))/(sum([rnaD[i][g] for g in nrcG]))) for i in ss],correction=True,alternative='less'))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ec386aa",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
